### DISCUSSION TITLE
Should there be a matrix check in `ggml_transpose`?

### DISCUSSION THREAD

**Discussion by ShriKaranHanda** [ggml-org/ggml#757] [Q&A] [answered]
I apologize in advance for any mistakes in understanding I may have! Still very new to tensors and ML frameworks.

Based on my understanding, `ggml_tranpose` only applies to 2D tensors (matrices). For tensors of higher dimension, `ggml_permute` should be used.

- Is my understanding correct?
- If it is correct, shouldn't there be a check in [`ggml_tranpose`](https://github.com/ggerganov/ggml/blob/fb558f78d905f85c54813602649ddd628ffe0f3a/src/ggml.c#L1956) for whether the input tensor is 2D or not? It seems to me that if it isn't 2D, then the `ggml_tranpose` function doesn't make any sense right?

Thanks!

**Comment by slaren** [comment, marked as answer]
`ggml_tranpose` swaps the first two dimensions of the tensor. That is usually fine even if the tensor has a higher dimension than 2, since the other dimensions are usually the batch size. So what you end with is a batch of transposed matrices, which is generally what you want. As you say, `ggml_permute` can be used for more advanced transposes that apply to any dimensions.
